### PR TITLE
Rename CONCAT() and STRINGIFY() macros

### PR DESCRIPTION
**Pull request by palchak-google**
Most of Zephyr's macros are prefix with with "Z_". This helps them avoid conflicting with macros from other codebases that may be used. However, the `CONCAT()` and `STRINGIFY()` macros do not have such prefixes, and both of these names are quite common. 

These two commits rename `CONCAT()` to `Z_CONCAT()` and `STRINGIFY()` to `Z_STRINGIFY()`.

### CODE REVIEW

**Comment by Nukersson** [comment]
Please provide reference to Zephyr's doc/requirements, where it mentioned, to use Z_ prefix for all macros. Additionally the docs must be updated also. Otherwise ok from my  side to Civetweb samples usage of the macros.

**Comment by palchak-google** [comment]
These two particular macros are not documented, so I'm not sure that there's any documentation to update. Also, I don't know if there's a hard requirement specified anywhere that all macros use the Z_ prefix. This change is just to maintain more consistency with the rest of the codebase and also to eliminate possible conflicts.

**Comment by nordic-krch** [comment]
`z_` prefix is used for non-public, internal APIs though i cannot find reference in the documentation (@nashif do you recall where it is put?) but i recall issues like #19497 were issuer stated: `The expectation is that the naming of functions, macros, and globals for logging follow the Zephyr coding standard that private APIs are prefixed with "z_".`

**Review by pfalcon — APPROVED** [review]
I'm definitely +1 on this, and sadly couldn't get it in on the previous try: #12935, #13426. Hopefully will be better this time.

File: subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5.h [hunk 1]
```diff
@@ -12,7 +12,7 @@
  * in radio_nrf5_ppi.h by the preprocessor.
  */
 #define EVENT_TIMER_ID 0
-#define EVENT_TIMER    _CONCAT(NRF_TIMER, EVENT_TIMER_ID)
+#define EVENT_TIMER    Z_CONCAT(NRF_TIMER, EVENT_TIMER_ID)
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5.h:15]: approved.

**Review by cvinayak — APPROVED** [review]
Approved changes in `subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5.h`

File: include/fs/fs.h [hunk 2]
```diff
@@ -209,7 +209,7 @@ struct fs_statvfs {
  * @brief The name under which a zephyr,fstab entry mount structure is
  * defined.
  */
-#define FS_FSTAB_ENTRY(node_id) _CONCAT(z_fsmp_, node_id)
+#define FS_FSTAB_ENTRY(node_id) Z_CONCAT(z_fsmp_, node_id)
```
Comment (de-nordic) [inline include/fs/fs.h:212]: +1, I am OK with the change.

**Review by de-nordic — APPROVED** [review]
OK with changes to the VFS.

**Review by nashif — CHANGES_REQUESTED** [review]
Z_ is for private macros and functions, the prefix came to replace plain _ (underscore) and to solve one issue with coding guidelines. Need to come up with a different prefix for this. Z_ usage is unfortunately a mess and this will just add fuel to the fire.

**Comment by pfalcon** [comment]
@nashif 

> Z_ is for private macros and functions, the prefix came to replace plain _ (underscore) and to solve one issue with coding guidelines. Need to come up with a different prefix for this. 

What's unprivate about CONCAT() and STRINGIFY()? Those are private macros for internal usage by Zephyr code modules, which just happened to be named like that because "everyone does it like that". They are not part of public API  or anything like that. Any more or less independent project which would need such facilities would define their own versions, not rely on Zephyr to provide them.

> Z_ usage is unfortunately a mess and this will just add fuel to the fire. 

As was mentioned at the previous time this issue was raised (see ticket references in my previous comments), this solves a real problem that "everyone" uses (defines) macros named like that, so it leads to conflicts and useless patching when trying to port existing software to Zephyr.

There're issues ("mess") with Zephyr API naming conventions, specifically the fact that Zephyr doesn't namespace (prefix) **all** of its API. But that's an orthogonal issue to finally at least consistently prefix all **private** symbols.

**Comment by dleach02** [comment]
Needs to be rebased and conflicts resolved

**Comment by cfriedt** [comment]
> Needs to be rebased and conflicts resolved

I think the fundamental issue is about whether the `CONCAT()` and `STRINGIFY()` macros are so conventional in the C programming language that they are de-facto standard.

I don't see anything wrong with making this change per se, but if not for this, then why not for `ARRAY_SIZE()` or `CONTAINER_OF()`? What about `MAX()` and `MIN()`? 

These are convenience macros that are nearly always used as they are, and namespacing them might be more trouble than it's worth.

@palchak-google - aside from the philosophical aspect, I would like to know why this change is being made. Are there any concrete instances where using these macros is causing a namespace collision?

**Comment by dkalowsk** [comment]
I can point to libraries that our products use which have headers with macro leakage similar in names.  While it is easy to discover and mitigate these kinds of issues, it's not a fun job.  I understand the argument against the Z_ prefix usage, and I get the rational for wanting to clean up the naming convention.  I don't believe the author is particularly wed to using the Z_ prefix, any prefix would suffice.  Unfortunately I don't see any recommended or proposed prefix within the PR comments, which makes this a block with no end.  How can the author move forward here?

I'm in favor of see this PR move forward rather than have it hang around, go stale, and allow the macro collision to continue to cause problems.

**Comment by cfriedt** [comment]
> I can point to libraries that our products use which have headers with macro leakage similar in names. While it is easy to discover and mitigate these kinds of issues, it's not a fun job. I understand the argument against the Z_ prefix usage, and I

I get that argument, but in that case, I think the counter argument is about which body of code should be forced to use a namespace.

I would consider Zephyr to be the more generic body of code that likely has the more conventional interpretation of the macros above. On the opposite end, I would guess that the 3rd party library or product is a less generic body of code and has a less conventional interpretation of the above macros.

I think the reason that @nashif does not want to accept this change is that he feels the less generic macro with a less conventional interpretation should be forced to namespace rather than forcing the more generic / more conventional to namespace.

Maybe that take is wrong though. Can someone clarify?

**Comment by dkalowsk** [comment]
> I get that argument, but in that case, I think the counter argument is about which body of code should be forced to use a namespace.

As macros are always public and global to a C compiler, the smart and safe practice is to always namespace them.  Otherwise you will run into issues like the long standing Visual Studio `MIN` and `MAX` collision issue that continues to impact developers, and results in things like the `NOMINMAX` define to work around it (every Windows developer has a story about this one biting them once).

If Z_ is not the prefix to use, I'll propose two alternatives: ZEPHYR_ or ZMACRO_ so this PR can move forward.

**Comment by cfriedt** [comment]
> If Z_ is not the prefix to use, I'll propose two alternatives: ZEPHYR_ or ZMACRO_ so this PR can move forward.

I'm ok with the `Z_` prefix, and I am actually fine approving this.

I think it really boils down to disrupting muscle-memory for a lot of devs who are accustomed to using the non-namespaced macros.

As much as I feel like it's just a clash of personal preference and opinions and style, and is not really a good use of time, this is precisely the reason that I think we have a TSC, and in this case, I would suggest that this be tagged TSC and have someone table this next week.

I will approve, but I'm not going to argue for this with the TSC - will anyone else volunteer to do that?

**Review by cfriedt — APPROVED** [review]

**Comment by dleach02** [comment]
I agree with @dkalowsk. It isn't helpful to apply a block indicating the z_ can't be used without providing the author an alternative that would pass muster for the block. @nashif, can you recommend a change that would gain your approval?

**Review by dleach02 — APPROVED** [review]

**Comment by dkalowsk** [comment]
@cfriedt good suggestion on the TSC.  If neither of the previous champions of the PR can attend I can try.

@pfalcon or @palchak-google I know your PR has been sitting around, but either of you able to represent the PR in front of the TSC?

**Comment by palchak-google** [comment]
What does representing this PR at the TSC entail?

**Comment by nashif** [comment]
> I think the reason that @nashif does not want to accept this change is that he feels the less generic macro with a less conventional interpretation should be forced to namespace rather than forcing the more generic / more conventional to namespace.

I am not against renaming this, rename it to anything, just do not use Z_, use something else. Is Z_ really the only way out here? You want options? What about UTIL_ for example?

> Anything else is a de-facto "public" macro, regardless of whether it's documented or not. An undocumented macro causes just as many name collision problems as a documented one.

yes, this is the point, it is public, renaming it to Z_ makes it count as private, i.e. it can be changed anytime without further notice.

Then there is that...
> I don't see anything wrong with making this change per se, but if so, then why not for `ARRAY_SIZE()` or `CONTAINER_OF()`? What about `MAX()` and `MIN()`?

Today we change STRINGIFY, tomorrow someone else has a conflict and this will not stop, but this is a different problem. 

> What does representing this PR at the TSC entail?

Just join the TSC and bring this as an open or mark the PR with TSC label and join :)

**Pull request closed without merging**